File: frontend/src/lib/events.ts
import { mockEvents, type Event } from '@/data/mockData'
import { supabase } from '@/lib/supabase'

interface EventRecord {
  id: string
  title: string
  summary: string
  description: string
  event_date: string
  start_time: string
  end_time: string | null
  location: string
  category: string
  tags: string[]
  confidence_tags: string[]
  organizer: string
  image_url: string | null
  why_recommended: string | null
  external_rsvp_url: string | null
  rsvp_count: number
}

function formatEventTime(startTime: string, endTime: string | null) {
  return endTime ? `${startTime} – ${endTime}` : startTime
}

export function mapEventRecord(record: EventRecord): Event {
  return {
    id: record.id,
    title: record.title,
    summary: record.summary,
    description: record.description,
    date: record.event_date,
    time: formatEventTime(record.start_time, record.end_time),
    location: record.location,
    category: record.category,
    tags: record.tags,
    confidenceTags: record.confidence_tags,
    imageUrl: record.image_url ?? '',
    rsvpCount: record.rsvp_count,
    organizer: record.organizer,
    whyRecommended: record.why_recommended ?? undefined,
    externalRsvpUrl: record.external_rsvp_url,
  }
}

export async function getEvents(limit?: number, offset?: number): Promise<Event[]> {
  let query = supabase
    .from('events')
    .select(`
      *,
      event_rsvp_data (
        rsvp_url,
        rsvp_type,
        cost,
        notes,
        source_url
      )
    `, { count: 'exact' })
    .order('event_date', { ascending: true })
    .order('start_time', { ascending: true })

  if (limit) {
    query = query.limit(limit)
  }

  if (offset) {
    query = query.range(offset, offset + (limit || 50) - 1)
  }

  const { data, error } = await query

  if (error) {
    console.warn('Falling back to bundled mock events because Supabase events could not be loaded.', error)
    return mockEvents
  }

  if (!data?.length) {
    console.warn('Supabase returned no events. Falling back to bundled mock events.')
    return mockEvents
  }

  return data.map(record => {
    const event = mapEventRecord(record)
    if (record.event_rsvp_data && record.event_rsvp_data.length > 0) {
      const rsvp = record.event_rsvp_data[0]
      event.rsvpData = {
        rsvpUrl: rsvp.rsvp_url,
        rsvpType: rsvp.rsvp_type,
        cost: rsvp.cost,
        notes: rsvp.notes,
        sourceUrl: rsvp.source_url,
      }
    }
    return event
  })
}

export async function getEventById(eventId: string): Promise<Event | null> {
  const { data, error } = await supabase.from('events').select('*').eq('id', eventId).maybeSingle()

  if (error) {
    console.warn(`Falling back to bundled mock event for ${eventId}.`, error)
    return mockEvents.find((event) => event.id === eventId) ?? null
  }

  if (!data) {
    return mockEvents.find((event) => event.id === eventId) ?? null
  }

  return mapEventRecord(data)
}

File: backend/app/services/events.py
from fastapi import HTTPException, status

from app.db.supabase import get_supabase
from app.models.events import EVENTS
from app.schemas.events import EventOut


def _format_time(start_time: str, end_time: str | None) -> str:
    return f"{start_time} - {end_time}" if end_time else start_time


def _map_event(record: dict) -> EventOut:
    return EventOut(
        id=record["id"],
        title=record["title"],
        summary=record["summary"],
        description=record["description"],
        date=str(record["event_date"]),
        time=_format_time(record["start_time"], record.get("end_time")),
        location=record["location"],
        category=record["category"],
        tags=record.get("tags", []),
        confidenceTags=record.get("confidence_tags", []),
        imageUrl=record.get("image_url") or "",
        rsvpCount=record.get("rsvp_count", 0),
        organizer=record["organizer"],
        whyRecommended=record.get("why_recommended"),
        externalRsvpUrl=record.get("external_rsvp_url"),
    )


def _map_seed_event(record: dict) -> EventOut:
    return EventOut(**record)


def _list_events_from_supabase(client) -> list[EventOut]:
    response = client.table("events").select("*").order("event_date").order("start_time").execute()
    return [_map_event(record) for record in (response.data or [])]


def list_events() -> list[EventOut]:
    try:
        return _list_events_from_supabase(get_supabase())
    except Exception:
        return [_map_seed_event(record) for record in EVENTS]


def get_event(event_id: str) -> EventOut:
    try:
        response = get_supabase().table("events").select("*").eq("id", event_id).limit(1).execute()

        if response.data:
            return _map_event(response.data[0])
    except Exception:
        pass

    for record in EVENTS:
        if record["id"] == event_id:
            return _map_seed_event(record)

    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event not found")
